rama-http/src/headers/x_robots_tag_components/valid_date.rs

```diff
@@ -0,0 +1,373 @@
+use chrono::{DateTime, FixedOffset, NaiveDateTime, Utc};
+use rama_core::error::{ErrorContext, OpaqueError};
+use std::collections::HashMap;
+use std::fmt::{Display, Formatter};
+use std::ops::Deref;
+use std::str::FromStr;
+use std::sync::OnceLock;
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub(super) struct ValidDate(DateTime<Utc>);
+
+impl Deref for ValidDate {
+    type Target = DateTime<Utc>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<ValidDate> for DateTime<Utc> {
+    fn from(value: ValidDate) -> Self {
+        value.0
+    }
+}
+
+impl From<DateTime<Utc>> for ValidDate {
+    fn from(value: DateTime<Utc>) -> Self {
+        Self(value)
+    }
+}
+
+impl AsRef<DateTime<Utc>> for ValidDate {
+    fn as_ref(&self) -> &DateTime<Utc> {
+        &self.0
+    }
+}
+
+impl AsMut<DateTime<Utc>> for ValidDate {
+    fn as_mut(&mut self) -> &mut DateTime<Utc> {
+        &mut self.0
+    }
+}
+
+impl FromStr for ValidDate {
+    type Err = OpaqueError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(ValidDate(
+            DateTime::parse_from_rfc3339(s) // check ISO 8601
+                .or_else(|_| {
+                    DateTime::parse_from_rfc2822(s) // check RFC 822
+                        .or_else(|_| datetime_from_rfc_850(s))
+                    // check RFC 850
+                })
+                .with_context(|| "Failed to parse date")?
+                .with_timezone(&Utc),
+        ))
+    }
+}
+
+impl Display for ValidDate {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        writeln!(f, "{}", self.0)
+    }
+}
+
+fn datetime_from_rfc_850(s: &str) -> Result<DateTime<FixedOffset>, OpaqueError> {
+    let (naive_date_time, remainder) = NaiveDateTime::parse_and_remainder(s, "%A, %d-%b-%y %T")
+        .with_context(|| "failed to parse naive datetime")?;
+
+    let fixed_offset = offset_from_abbreviation(remainder)?;
+
+    Ok(DateTime::from_naive_utc_and_offset(
+        naive_date_time,
+        fixed_offset,
+    ))
+}
+
+fn offset_from_abbreviation(remainder: &str) -> Result<FixedOffset, OpaqueError> {
+    get_timezone_map()
+        .get(remainder.trim())
+        .ok_or_else(|| OpaqueError::from_display(format!("invalid abbreviation: {}", remainder)))?
+        .parse()
+        .with_context(|| "failed to parse timezone abbreviation")
+}
+
+static TIMEZONE_MAP: OnceLock<HashMap<&'static str, &'static str>> = OnceLock::new();
+
+fn get_timezone_map() -> &'static HashMap<&'static str, &'static str> {
+    TIMEZONE_MAP.get_or_init(|| {
+        let mut map = HashMap::new();
```

Comment:
code is gonna be cleaner and more efficient if you collect the hashmap from an array of tuples